azkaban-common/src/main/java/azkaban/version/JdbcVersionSetLoaderImpl.java

```diff
@@ -0,0 +1,217 @@
+            throw new IOException("Exception occurred while removing version set for " + versionSetMd5Hex, e);
+        }
+    }
+
+    /**
+     * This method first checks if the versionSetNum already exists in the local copy of the versionSets,
+     * otherwise, it insert the versionSet into the table and get the resulting versionSetNum.
+     */
+    @Override
+    public synchronized int getVersionSetNum(String versionSetJsonString, String versionSetMd5Hex) throws IOException {
+        if (this.md5ToVersionSet.containsKey(versionSetMd5Hex)) {
+            VersionSet versionSet = this.md5ToVersionSet.get(versionSetMd5Hex);
+            return versionSet.getVersionSetNum();
+        }
+        return insertAndGetVersionSetNum(versionSetJsonString, versionSetMd5Hex);
+    }
+
+    /**
+     * This method first check if the versionSet exists in the local copy of versionSets,
+     * otherwise, it gets the copy from the version_set table and updates the local copy.
+     */
+    @Override
+    public synchronized VersionSet getVersionSet(String versionSetMd5Hex) throws IOException {
+        if (this.md5ToVersionSet.containsKey(versionSetMd5Hex)) {
+            return this.md5ToVersionSet.get(versionSetMd5Hex);
+        }
+        final VersionSetHandler versionSetHandler = new VersionSetHandler();
+        try {
+            List<VersionSet> versionSets = this.dbOperator.query(SELECT_VSET_FROM_MD5, versionSetHandler, versionSetMd5Hex);
+            if (versionSets.size() > 1) {
+                throw new IOException("Expected only one VersionSet for versionSetMd5Hex " + versionSetMd5Hex);
+            }
+            if (versionSets.isEmpty()) {
+                logger.info("No VersionSet exists for versionSetMd5Hex " + versionSetMd5Hex);
+                return null;
+            }
+            VersionSet versionSet = versionSets.get(0);
+            this.numToVersionSet.put(versionSet.getVersionSetNum(), versionSet);
+            this.md5ToVersionSet.put(versionSetMd5Hex, versionSet);
+            return versionSet;
+        } catch (SQLException e) {
+            logger.error("Failed to execute: " + SELECT_VSET_FROM_MD5, e);
+            throw new IOException("Failed to fetch VersionSet for versionSetMd5Hex " + versionSetMd5Hex, e);
+        }
+    }
+
+    /**
+     * This method first check if the versionSet exists in the local copy of versionSets,
+     * otherwise, it gets the copy from the version_set table and updates the local copy.
+     */
+    @Override
+    public synchronized VersionSet getVersionSet(int versionSetNum) throws IOException {
+        if (this.numToVersionSet.containsKey(versionSetNum)) {
+            return this.numToVersionSet.get(versionSetNum);
+        }
+        final VersionSetHandler versionSetHandler = new VersionSetHandler();
+        try {
+            List<VersionSet> versionSets = this.dbOperator.query(SELECT_VSET_FROM_NUM, versionSetHandler, versionSetNum);
+            if (versionSets.size() > 1) {
```

Comment:
There is a common code in some get methods. Can you please extract common code in another method and use it everywhere?

Reply:
Common code is there with the exchange of num and Id, along with log messages, etc. Hence it will be better to keep it segregated